File: garnet/public/lib/inspect/health/health.cc
// Copyright 2019 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "health.h"

namespace inspect {

NodeHealth::NodeHealth(Node* parent_node)
    : health_node_(parent_node->CreateChild(kHealthNodeName)),
      health_status_(
          health_node_.CreateStringProperty("status", kHealthStartingUp)) {}

void NodeHealth::Ok() {
  health_message_.reset();
  health_status_.Set(kHealthOk);
}

void NodeHealth::StartingUp() {
  health_message_.reset();
  health_status_.Set(kHealthStartingUp);
}

void NodeHealth::StartingUp(const std::string& message) {
  health_status_.Set(kHealthStartingUp);
  SetMessage(message);
}

void NodeHealth::Unhealthy(const std::string& message) {
  health_status_.Set(kHealthUnhealthy);
  SetMessage(message);
}

void NodeHealth::SetStatus(const std::string& status,
                           const std::string& message) {
  health_status_.Set(status);
  SetMessage(message);
}

void NodeHealth::SetMessage(const std::string& message) {
  if (!health_message_) {
    health_message_ = health_node_.CreateStringProperty("message", "");
  }
  health_message_->Set(message);
}

}  // namespace inspect

File: src/security/scrutiny/plugins/src/engine/controller/health.rs
// Copyright 2020 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use {
    anyhow::Result,
    scrutiny::{model::controller::DataController, model::model::DataModel},
    scrutiny_utils::usage::UsageBuilder,
    serde_json::{json, value::Value},
    std::sync::Arc,
};

#[derive(Default)]
pub struct HealthController {}

/// The `HealthController` simply returns a ping. This is used to determine if
/// the service is alive and operating.
impl DataController for HealthController {
    fn query(&self, _: Arc<DataModel>, _: Value) -> Result<Value> {
        Ok(json!({"status" : "ok"}))
    }
    fn description(&self) -> String {
        "Health endpoint that always returns ok.".to_string()
    }
    fn usage(&self) -> String {
        UsageBuilder::new()
            .name("engine.health.status - Health endpoint")
            .summary("engine.health.status")
            .description(
                "Health endpoint that always returns ok. \
            This is used mostly by remote endpoints to determine if the \
            connection and the api is available",
            )
            .build()
    }
}
